docs/ConstPropagationPass.md

```diff
@@ -0,0 +1,266 @@
+sparse and we now support dense constant tensors only. We need additionallly
+check a dense constant tensor by using `IsFromDenseONNXConstantOp`.
+
+In the result pattern, to produce a ONNXConstantOp, we will add `lhs`
+and `rhs` at compile time, and emit an ONNXConstantOp. To minimize the
+memory footprint, **this ONNXConstant does not have a DenseElementsAttr**, but
+refers to an internal buffer where the real data is stored. DenseElementsAttrs
+will be added to only **the final ONNXConstantOps of the whold pass**,
+not to intermediate generated ONNXConstantOps.
+
+Function `CreateAddOfTwoConst` will do the addition at compile time and return
+an ONNXConstant.
+
+```
+def CreateAddOfTwoConst :
+   NativeCodeCall<"ConstPropElementwiseBinary<mlir::ONNXAddOp>($_builder, $0, $1, $2)">;
+```
+
+### Step 2: Prepare array buffers for inputs and result <a id="step2"></a>
+
+Function `CreateAddOfTwoConst` in the pattern calls
+`ConstPropElementwiseBinary` in [ConstProp.cpp](https://github.com/onnx/onnx-mlir/blob/master/src/Transform/ONNX/ConstProp.cpp) whose content is as follows.
+
+```c++
+template <typename ElementwiseBinaryOp>
+ONNXConstantOp ConstPropElementwiseBinary(
+    PatternRewriter &rewriter, Value replacingValue, Value lhs, Value rhs) {
+  Type elementType =
+      replacingValue.getType().cast<ShapedType>().getElementType();
+  ArrayRef<int64_t> lhsShape = lhs.getType().cast<ShapedType>().getShape();
+  ArrayRef<int64_t> rhsShape = rhs.getType().cast<ShapedType>().getShape();
+  ArrayRef<int64_t> outputShape =
+      replacingValue.getType().cast<ShapedType>().getShape();
+      
+  // Get lhs and rhs array buffers.
+  char *lhsArray = getArrayFromAttributeOrBuffer(rewriter, lhs.getDefiningOp());
+  char *rhsArray = getArrayFromAttributeOrBuffer(rewriter, rhs.getDefiningOp());
+
+  // Allocate a buffer for the result.
+  // Use maximum size (double or int64_t) to avoid the precision loss.
+  char *resArray =
+      allocateBufferFor(replacingValue.getType(), /*useMaxSize=*/true);
+      
+  // Do calculation on array buffers.
+  if (elementType.isa<FloatType>()) {
+    // Use double to avoid the precision loss during computation.
+    IterateConstPropElementwiseBinary<ElementwiseBinaryOp, double>(
+        lhsArray, rhsArray, lhsShape, rhsShape, resArray, outputShape);
+  } else if (elementType.isa<IntegerType>()) {
+    // Use int64_t to avoid the precision loss during computation.
+    IterateConstPropElementwiseBinary<ElementwiseBinaryOp, int64_t>(
+        lhsArray, rhsArray, lhsShape, rhsShape, resArray, outputShape);
+  } else
+    llvm_unreachable("Unknown data type");
+
+  // Construct a new ONNXConstantOp for the result array buffer.
+  // This ONNXConstantOp contains a buffer ID instead of a DenseElementsAttr.
+  ONNXConstantOp res =
+      createConstantOpAndStoreBufferPtr(rewriter, replacingValue, resArray);
+  
+  // Return the constant.
+  return res;
+}
+```
```

Comment:
What if we had an FMA, and we decided to create first a buffer for multiplying, and then a second buffer for adding, and then only associate the second buffer with the constant op. Would the first buffer "escape" the freeing?

Reply:
We need to free the first buffer manually. I added an explanation about this in the Remark section.